### PR TITLE
Reject invalidomain with wildcards

### PR DESCRIPTION
**Pull request by cyd01**
I propose an update to my previous PR: add reject from origin domain feature #375
In this PR, I propose a new feature to check origin smtp domain against a list that can contains joker characters (* and ?).

### CODE REVIEW

File: pkg/stringutil/utils.go [hunk 1]
```diff
@@ -74,3 +74,43 @@ func MakePathPrefixer(prefix string) func(string) string {
 		return prefix + path
 	}
 }
+
+// Test if a "s" string match a "p" pattern with joker (*, ?)
+func MatchWithJokers(p string, s string) bool {
```
Comment (jhillyerd) [inline pkg/stringutil/utils.go:79]: Let's use the more common term "wildcard" instead of "joker"

I'd prefer this function to have a couple table-driven tests to cover matching and non-matching scenarios for edge cases, such as:

* pattern contains `*`, `?`, `**`, `??`, `*?`, `?*`
* pattern starts with `*`, `?`, `**`, `??`, `*?`, `?*`
* pattern ends with `*`, `?`, `**`, `??`, `*?`, `?*`
* empty strings for pattern and string to be matched

File: pkg/stringutil/utils.go [hunk 2]
```diff
@@ -74,3 +74,43 @@ func MakePathPrefixer(prefix string) func(string) string {
...
+	runeInput := []rune(s)
+	runePattern := []rune(p)
+	lenInput := len(runeInput)
+	lenPattern := len(runePattern)
+	isMatchingMatrix := make([][]bool, lenInput+1)
+	for i := range isMatchingMatrix {
+		isMatchingMatrix[i] = make([]bool, lenPattern+1)
+	}
+	isMatchingMatrix[0][0] = true
+	for i := 1; i < lenInput; i++ {
```
Comment (jhillyerd) [inline pkg/stringutil/utils.go:89]: I think this loop can be eliminated as Go will initialize a slice of bool to false

File: pkg/policy/address.go [hunk 3]
```diff
@@ -117,7 +117,15 @@ func (a *Addressing) ShouldStoreDomain(domain string) bool {
...
 	domain = strings.ToLower(domain)
-	return !stringutil.SliceContains(a.Config.SMTP.RejectOriginDomains, domain)
+	if len(a.Config.SMTP.RejectOriginDomains) > 0 {
+		for _, d := range a.Config.SMTP.RejectOriginDomains {
+			if stringutil.MatchWithJokers(d, domain) {
+				return false
+			}
+		}
+	}
+	//return !stringutil.SliceContains(a.Config.SMTP.RejectOriginDomains, domain)
```
Comment (jhillyerd) [inline pkg/policy/address.go:127]: remove line

**Review by jhillyerd — CHANGES_REQUESTED** [review]

**Review by jhillyerd — APPROVED** [review]

**Comment by jhillyerd** [comment]
Thanks!  I've opened #425 for the Docker check failure
